File: 3611-construct-the-minimum-bitwise-array-ii/construct-the-minimum-bitwise-array-ii.cpp
class Solution {
public:
    vector<int> minBitwiseArray(vector<int>& nums) {
        int n=nums.size();
        vector<int>ans(n,-1);
        for(int i=0;i<n;i++){
            int num=nums[i];
            int j=0;
            for(;j<32;j++){
                if(!(num & (1<<j))){
                    break;
                } else {
                    num=num ^ (1<<j);
                }
            }

            for(int k=0;k<(1<<j);k++){
                int curr=num | k;
                if((curr | (curr+1))==nums[i]){
                    ans[i]=curr;
                    break;
                }
            }
        }
        return ans;
    }
};

File: 3611-construct-the-minimum-bitwise-array-ii/construct-the-minimum-bitwise-array-ii.java
class Solution {

    public static int calculate(int x) {
        int shiftBy=0;
        while((x & (1<<shiftBy))!=0 && shiftBy<32){
            shiftBy++;
        }
       return x ^ (1<<(shiftBy-1));
    }

    public int[] minBitwiseArray(List<Integer> nums) {
        int n = nums.size();
        int[] ans = new int[n];
        Arrays.fill(ans, Integer.MAX_VALUE);
        for (int i = 0; i < n; i++) {
            if ((nums.get(i) & 1) == 1) {
                // making last bit 0
                Integer poss1 = nums.get(i) & (~1);

                // second solution
                Integer poss2 = calculate(nums.get(i));

                System.out.println(nums.get(i)+" "+ poss1+ " "+poss2);

                if ((poss1 | (poss1 + 1)) == nums.get(i)) {
                    ans[i] = Math.min(ans[i], poss1);
                }

                if ((poss2 | (poss2 + 1)) == nums.get(i)) {
                    ans[i] = Math.min(ans[i], poss2);
                }

            } else {
                ans[i] = -1;
            }
        }
        return ans;
    }

}